packages/nuxt/src/app/composables/loading-indicator.ts

```diff
@@ -0,0 +1,135 @@
+  duration: number
+  /** @default 200 */
+  throttle: number
+}
+
+function _increase (progress: Ref<number>, num: number) {
+  progress.value = Math.min(100, progress.value + num)
+}
+
+function _hide (isLoading: Ref<boolean>, progress: Ref<number>) {
+  if (import.meta.client) {
+    setTimeout(() => {
+      isLoading.value = false
+      setTimeout(() => { progress.value = 0 }, 400)
+    }, 500)
+  }
+}
+
+export type LoadingIndicator = {
+  _cleanup: () => void
+  progress: Ref<number>
+  isLoading: Ref<boolean>
+  start: () => void
+  finish: () => void
+  clear: () => void
+}
+
+function createLoadingIndicator (opts: Partial<LoadingIndicatorOpts> = {}) {
+  const { duration = 2000, throttle = 200 } = opts
+  const nuxtApp = useNuxtApp()
+  const progress = ref(0)
+  const isLoading = ref(false)
+  const step = computed(() => 10000 / duration)
+
+  let _timer: any = null
+  let _throttle: any = null
+
+  function start () {
+    if (nuxtApp.isHydrating) {
+      return
+    }
+    clear()
+    progress.value = 0
+    if (throttle && import.meta.client) {
+      _throttle = setTimeout(() => {
+        isLoading.value = true
+        _startTimer()
+      }, throttle)
+    } else {
+      isLoading.value = true
+      _startTimer()
+    }
+  }
+
+  function finish () {
+    progress.value = 100
+    clear()
+    _hide(isLoading, progress)
+  }
+
+  function clear () {
+    clearInterval(_timer)
+    clearTimeout(_throttle)
+    _timer = null
+    _throttle = null
+  }
+
+  function _startTimer () {
+    if (import.meta.client) {
+      _timer = setInterval(() => { _increase(progress, step.value) }, 100)
+    }
+  }
+
+  let _cleanup = () => {}
+  if (import.meta.client) {
+    const unsubLoadingStartHook = nuxtApp.hook('page:loading:start', () => {
+      start()
+    })
+    const unsubLoadingFinishHook = nuxtApp.hook('page:loading:end', () => {
+      finish()
+    })
+    const unsubError = nuxtApp.hook('vue:error', finish)
+
+    _cleanup = () => {
+      unsubError()
+      unsubLoadingStartHook()
+      unsubLoadingFinishHook()
+      clear()
+    }
+  }
+
+  return {
+    _cleanup,
+    progress,
+    isLoading,
+    start,
+    finish,
+    clear
+  }
+}
+
+/**
+ * composable to handle the loading state of the page
+ */
+export function useLoadingIndicator (opts: Partial<LoadingIndicatorOpts> = {}) {
+  const nuxtApp = useNuxtApp()
+
+  // Initialise global loading indicator if it doesn't exist already
+  const indicator = nuxtApp._loadingIndicator = nuxtApp._loadingIndicator || createLoadingIndicator(opts)
+  if (import.meta.client && getCurrentScope()) {
+    nuxtApp._loadingIndicatorDeps = nuxtApp._loadingIndicatorDeps || 0
+    nuxtApp._loadingIndicatorDeps++
+    onScopeDispose(() => {
+      nuxtApp._loadingIndicatorDeps!--
+      if (nuxtApp._loadingIndicatorDeps === 0) {
+        indicator._cleanup()
+        delete nuxtApp._loadingIndicator
+      }
+    })
+  }
+
+  return {
+    progress: indicator.progress,
```

Comment:
Perf: We could avoid a (second) object creation each time this method is being called by directly bypassing `indicator`. I see that it hides `_cleanup` if it is essential to remove access from users we could benefit from a symbol but really it seems okay and also nice escape hatch..